code/modules/mob/living/carbon/xenomorph/Powers.dm

```diff
@@ -1106,16 +1106,70 @@
 			break
 
 		if(stagger)
-			to_chat(src, "<span class='xenowarning'>Your muscles fail to respond as you try to shake up the shock!</span>")
+			to_chat(src, "<span class='xenowarning'>You try to salvage [target]'s essence but are staggered!</span>")
 			break
 
-		if(target.plasma_stored < amount)
-			amount = target.plasma_stored //Just take it all.
+		face_atom(target)
+		animation_attack_on(target)
+		playsound(target, pick('sound/voice/alien_drool1.ogg', 'sound/voice/alien_drool2.ogg'), 15, 1)
+
+		round_statistics.drone_salvage_essence++
+
+		var/absorbed_plasma = round(target.plasma_stored * DRONE_SALVAGE_PLASMA_MULTIPLIER)
+		var/absorbed_evolution = round(target.evolution_stored * DRONE_SALVAGE_EVOLUTION_MULTIPLIER)
+		var/absorbed_upgrade = round(target.upgrade_stored * DRONE_SALVAGE_UPGRADE_MULTIPLIER)
+		var/excess_plasma
+
+		if(absorbed_plasma > (xeno_caste.plasma_max - plasma_stored) ) //Amount exceeding remaining plasma capacity absorbed as health
+			excess_plasma = absorbed_plasma - (xeno_caste.plasma_max - plasma_stored)
+			adjustFireLoss(-CLAMP( excess_plasma, 0, getFireLoss() ) )
+			adjustBruteLoss(-CLAMP( excess_plasma, 0, getBruteLoss() ) )
+
+		absorbed_plasma = absorbed_plasma - excess_plasma
+		gain_plasma(absorbed_plasma)
+
+		var/list/upgrade_list = list()
+		var/list/evolution_list = list()
+		var/mob/living/carbon/Xenomorph/X
+
+		for(var/i in GLOB.alive_xeno_list)
```

Comment:
Can you put this expensive `for` loop outside of the `while`, for when the absorption operation is over?